packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -871,6 +874,97 @@ impl<'m> RunnerImpl<'m> {
         Ok(())
     }
 
+    /// TODO
+    fn handle_task_msg(
+        &mut self,
+        mv8: &'m MiniV8,
+        sim_id: SimulationShortId,
+        msg: RunnerTaskMsg,
+        outbound_sender: &UnboundedSender<OutboundFromRunnerMsg>,
+    ) -> Result<()> {
+        tracing::debug!("Starting state interim sync before running task");
+        // TODO: Move JS part of sync into `run_task` function in JS for better performance.
+        self.state_interim_sync(mv8, sim_id, &msg.shared_store)?;
+
+        tracing::debug!("Setting up run_task function call");
+
+        let (payload, wrapper) = msg
+            .payload
+            .extract_inner_msg_with_wrapper()
+            .map_err(|err| {
+                Error::from(format!("Failed to extract the inner task message: {err}"))
+            })?;
+        let payload_str = mv8::Value::String(mv8.create_string(&serde_json::to_string(&payload)?));
+
+        let context = msg.shared_store.context().clone();
+        let shared_stores = match msg.shared_store.state {
+            SharedState::None | SharedState::Write(_) | SharedState::Read(_) => {
+                // Run the task on all groups
+                vec![(None, msg.shared_store)]
+            }
+            SharedState::Partial(partial_shared_state) => partial_shared_state
+                .split_into_individual_per_group()
+                .into_iter()
+                .map(|shared_state| {
+                    debug_assert_eq!(shared_state.indices().len(), 1);
+                    let group_index = shared_state.indices()[0];
+                    (
+                        Some(group_index),
+                        TaskSharedStore::new(SharedState::Partial(shared_state), context.clone()),
```

Comment:
I don't think, the context has to be cloned as it's already cloned in  L899.

Reply:
It needs to be cloned in the iterator, and if we don't clone it at the start, then we get a problem of a partial move with a borrow. (Because context() is a method, so it's not smart enough to deconstruct the struct)

Reply:
```rust
error[E0505]: cannot move out of `msg.shared_store.state.0` because it is borrowed
   --> src/worker/runner/javascript/mod.rs:905:34
    |
899 |         let context = msg.shared_store.context();
    |                       -------------------------- borrow of `msg.shared_store` occurs here
...
905 |             SharedState::Partial(partial_shared_state) => partial_shared_state
    |                                  ^^^^^^^^^^^^^^^^^^^^ move out of `msg.shared_store.state.0` occurs here
...
913 |                         TaskSharedStore::new(SharedState::Partial(shared_state), context.clone()),
    |                                                                                  ------- borrow later captured here by closure
```

Reply:
My bad.